draft-ietf-rtcweb-jsep.xml

```diff
@@ -2359,11 +2359,18 @@ candidate:1 1 UDP 1694498815 192.0.2.33 10000 typ host
             be generated according to the same rules as for an initial
             offer.</t>
 
-            <t>Each "m=" and c=" line MUST be filled in with the port,
-            protocol, and address of the default candidate for the m=
+            <t>Each "m=" and c=" line MUST be filled in with the port
+            and address of the default candidate for the m=
```

Comment:
This should be address and port of the default candidate if it matches the protocol in the m= line or IN IP4 0.0.0.0 and port 9, if protocol does not match the default candidate. Otherwise it adds an additional candidate or causes the mismatch.

Reply:
So just to make sure I understand, let me make an example. Say the offer came from a server with a public IP and only does ice light and it offers an IP of 1.2.3.4 with port 2000 for TCP and 3000 for UDP. And lets say UDP is blocked and does not work. You are concerned that the m line in answer could end up IP and port mapped to the TCP flow, but the protocol would say UDP not TCP and that would be really weird? I'm just not following the concern here so if you could spell it out a bit more, that would be great. Thanks

Reply:
My concern here is that if server on public IP which implements *full* ICE sends an offer with IP of 1.2.3.4 with port 2000 for TCP and 3000 for UDP. Let's say UDP is blocked or disabled on the client, so server get's back an answer with only TCP candidate with IP of 2.3.4.5 and port 4000 but m and c lines would have UDP protocol with IP 2.3.4.5 and port 4000. According to ice-sip-sdp this will be treated as an additional candidate by the server, with server sending STUN message over UDP to IP 2.3.4.5 port 4000. With older ICE implementations this will be treated as ICE mismatch. Finally, this is just plain weird and misleading.

What I am suggesting is that if none of the candidates in the answer match the protocol in the offer, send an answer with protocol from the offer, IN IP4 0.0.0.0 and port 9. So, in our example client answer will have only a single TCP candidate, IN IP4 0.0.0.0 in the c= line and port 9 in the m line. The server, when it receives this answer will know that information is a place holder and will not treat it as an additional candidate or mismatch.

Reply:
Hmm - you could be right but can you point me at where in  ice-sip-sdp it takes something that arrives in the m= line and forms a new candidate out of it? That would seem wrong in that when doing ice, only candidates that are in candidate lines should be used. I get the m= line would indicate the default candidate and how that might cause some confusion here but I don't think ICE should change for it to form a new candidate as that seems like it will have bad backwards compatible interoperability problems.

Reply:
This is coming from ice-sip-sdp section 3.2.5 (https://tools.ietf.org/html/draft-ietf-mmusic-ice-sip-sdp-24#section-3.2.5).

The idea was that if firewall is modifying the address in the c= and m= line to an address that does not match any of the candidates, it is a better to run a connectivity check against the firewall provided address together with all the ICE candidates vs declaring the mismatch,  ignoring all the ICE candidates and only using the address in c= and m= line (as specified in RFC5245).